src/cpp/src/visual_language/vision_encoder.hpp

```diff
@@ -120,51 +86,42 @@ class VisionEncoder {
     /// instead of the config obtained in constructors.
     /// @return Resulting embeddings for the resized source image and
     /// its slices.
-    EncodedImage encode(
-        const ov::Tensor& image, const ov::AnyMap& config_map
-    );
+    EncodedImage encode(const ov::Tensor& image, const ov::AnyMap& config_map = {});
 
-    /// @brief Compute embeddings of an image given
-    /// ProcessorConfig members.
-    /// @param image An image to infer embeddings for. Image shape must be
-    /// [1CHW]. Only batch 1 is supported.
-    /// @param ...properties A config or its members values to follow
-    /// instead of the config obtained in constructors.
-    /// @return Resulting embeddings for the resized source image and
-    /// its slices.
-    template <typename... Properties>
-    util::EnableIfAllStringAny<EncodedImage, Properties...> encode(
-        const ov::Tensor& image,
-        Properties&&... properties
-    ) {
-        return encode(
-            image, AnyMap{std::forward<Properties>(properties)...}
-        );
-    }
+    ProcessorConfig get_processor_config() const;
 
 private:
-    EncodedImage encode_minicpm(
-        const ov::Tensor& image, const ProcessorConfig& config
-    );
-
-    EncodedImage encode_llava(
-        const ov::Tensor& image, const ProcessorConfig& config
-    );
-
-    EncodedImage encode_llava_next(
-        const ov::Tensor& image, const ProcessorConfig& config
-    );
-
-    EncodedImage encode_internvl(
-        const ov::Tensor& image, const ProcessorConfig& config
-    );
-
-    EncodedImage encode_phi3_v(
-        const ov::Tensor& image, const ProcessorConfig& config
-    );
-
-    EncodedImage encode_qwen2vl(
-        const ov::Tensor& image, const ProcessorConfig& config
-    );
+    // base class for model-specific vision encoders
+    class IVisionEncoder {
+    public:
+        IVisionEncoder(const std::filesystem::path& model_dir, const std::string& device, const ov::AnyMap properties);
```

Comment:
`IVisionEncoder` takes almost all the tasks which were performed by `VisionEncoder` earlier. `VisionEncoder` only constructs the actual VisionEncoder* and forwards the calls now. Removing `VisionEncoder` and using a pointer to `IVisionEncoder` would do the same. I think you should drop `VisionEncoder` class and add factory functions instead of `VisionEncoder` constructors.